### PR TITLE
Add regularization to GP hyperparam optimization

### LINKED ISSUE #23026: Add regularization to GP hyperparam optimization
## Reason
Regularization by including prior distributions over the hyperparams of GP in the loss function aids in the optimization

## Design
The log of hyperparam priors are all fixed to a Normal distribution centered at zero. An optional standard deviation value is requested. By default, this value is set to 1000.0 and this shouldn't have any impact on the existing behavior and results. It's when the standard deviation value is lowered the effect of regularization kicks in.

## Impact
No anticipated impacts to existing objects.

### PR DESCRIPTION
**Pull request by somu15**
closes #23026

### CODE REVIEW

File: modules/stochastic_tools/test/tests/surrogates/gaussian_process/GP_Matern_half_int_tuned_adam.i [hunk 1]
```diff
@@ -101,6 +101,7 @@
     iter_adam = 1000
     batch_size = 20
     learning_rate_adam = 0.005
+    prior_std =1000.0
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/surrogates/gaussian_process/GP_Matern_half_int_tuned_adam.i:104]: I would just remove this from the input, since it is the default value

File: modules/stochastic_tools/test/tests/surrogates/gaussian_process/GP_Matern_half_int_tuned_adam.i [hunk 2]
```diff
@@ -126,6 +127,6 @@
   [out]
     type = CSV
     execute_on = FINAL
-    file_base = 'GP_Matern_half_int_tuned_adam'
+    file_base ='GP_Matern_half_int_tuned_adam'
```
Comment (zachmprince) [inline modules/stochastic_tools/test/tests/surrogates/gaussian_process/GP_Matern_half_int_tuned_adam.i:130]: Revert this change.

File: modules/stochastic_tools/doc/content/source/surrogates/GaussianProcessTrainer.md [hunk 3]
```diff
@@ -144,7 +144,7 @@ The following options are available for tuning the hyperparameters:
 
 - +Adaptive moment estimation (Adam)+
 
-  Relies on the psuedocode provied in [!cite](kingma2014adam). Adam permits stochastic optimization, wherein, a batch of the training data can be randomly chosen at each iteration.
+  Relies on the psuedocode provied in [!cite](kingma2014adam). Adam permits stochastic optimization, wherein, a batch of the training data can be randomly chosen at each iteration. Additionally, there is an option to regularize the hyper-params in the loss function by specifying a standard deviation value. This value corresponds to a Normal distribution centered at zero over the log of the hyper-param. By default, the standard deviation is set to a 1000 and this would have no effect on the optimization. As its value reduces, the regularization effects kicks-in.
```
Comment (zachmprince) [inline modules/stochastic_tools/doc/content/source/surrogates/GaussianProcessTrainer.md:147]: Could you comment on when regularization should be considered?

File: modules/stochastic_tools/include/utils/GaussianProcessHandler.h [hunk 4]
```diff
@@ -153,13 +156,19 @@ class GaussianProcessHandler
                            unsigned int iter,
                            const unsigned int & batch_size,
                            const Real & learning_rate,
-                           const bool & verbose);
+                           const bool & verbose,
+                           const Real & p_std);
```
Comment (zachmprince) [inline modules/stochastic_tools/include/utils/GaussianProcessHandler.h:160]: I would switch the order of these two arguments. Debugging or verbosity arguments are generally last.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C [hunk 5]
```diff
@@ -44,6 +44,11 @@ GaussianProcessTrainer::validParams()
       "tao_options", "", "Command line options for PETSc/TAO hyperparameter optimization");
   params.addParam<bool>(
       "show_optimization_details", false, "Switch to show TAO or Adam solver results");
+  params.addRangeCheckedParam<Real>(
+      "prior_std",
+      1000.0,
+      "prior_std>0.0",
+      "The standard deviation of a Normal prior over the log of hyper-params");
```
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:51]: Is this option valid for TAO optimization? If not, I'd recommend throwing an error if it is specified and not used.

File: modules/stochastic_tools/src/utils/GaussianProcessHandler.C [hunk 6]
```diff
@@ -361,11 +369,16 @@ GaussianProcessHandler::getLossAdam(RealEigenMatrix & inputs, RealEigenMatrix &
   log_likelihood += -std::log(_K.determinant());
   log_likelihood -= _batch_size * std::log(2 * M_PI);
   log_likelihood = -log_likelihood / 2;
+  for (unsigned int ii = 0; ii < _num_tunable; ++ii)
+    log_likelihood += -std::log(p_std) - std::log(2.0 * M_PI) / 2.0 -
+                      Utility::pow<2>((std::log(theta(ii)) / p_std)) / 2.0;
```
Comment (zachmprince) [inline modules/stochastic_tools/src/utils/GaussianProcessHandler.C:374]: Are there other types of regularization that could be implemented in the future? If so, I'm thinking we could have a function that adds the regularization: something like `getRegularization(theta)`. This would have the logic to switch between regularization types.

**Review by zachmprince — CHANGES_REQUESTED** [review]
Some general comments and questions.

File: modules/stochastic_tools/src/utils/GaussianProcessHandler.C (hunk 6 above)
Comment (somu15) [reply modules/stochastic_tools/src/utils/GaussianProcessHandler.C:374]: You're right. There are other types of regularizations. But all methods don't seem to work in the same way in terms of modifying the loss function. But for the Bayesian regularization, at least, I will generalize this by adding the `getRegularization()` function.

File: modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C (hunk 5 above)
Comment (somu15) [reply modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C:51]: I don't know the validity of this to TAO because TAO seems to formulate its own loss function. Sure, I will add an errror.

File: modules/stochastic_tools/src/utils/GaussianProcessHandler.C (hunk 6 above)
Comment (somu15) [reply modules/stochastic_tools/src/utils/GaussianProcessHandler.C:374]: These two are useful resources on priors for the GP hyper-params: https://mc-stan.org/docs/2_22/stan-users-guide/fit-gp-section.html 

https://www.tensorflow.org/probability/examples/Gaussian_Process_Regression_In_TFP

**Comment by somu15** [comment]
A small change is needed to `ActiveLearningGaussianProcess` once https://github.com/idaholab/moose/pull/22282 is merged. I just need to add an error check.

**Comment by somu15** [comment]
> A small change is needed to `ActiveLearningGaussianProcess` once #22282 is merged. I just need to add an error check.

This is done now.
